spring-batch-docs/asciidoc/spring-batch-integration.adoc

```diff
@@ -935,19 +935,103 @@ public ChunkProcessorChunkHandler chunkProcessorChunkHandler() {
 link:$$http://docs.spring.io/spring-batch/reference/html/scalability.html#remoteChunking$$[Remote Chunking].
 
+Starting from version 4.1, Spring Batch Integration provides the `@EnableBatchIntegration`
+annotation that can be used to simplify remote chunking setup. This annotation provides
+two beans that can be autowired in the application context:
+
+* `RemoteChunkingMasterStepBuilderFactory`: used to configure the master step
+* `RemoteChunkingWorkerBuilder`: used to configure the remote worker integration flow
+
+These APIs will take care of configuring a number of components as described in the following diagram:
+
+.Remote Chunking Configuration
+image::{batch-asciidoc}images/remote-chunking-config.png[Remote Chunking Configuration, scaledwidth="80%"]
+
+On the master side, the `RemoteChunkingMasterStepBuilderFactory` allows you to
+configure a master step by declaring:
+
+* the item reader to read items ans send them to workers
+* the output channel ("Outgoing requests") to send requests to workers
+* the input channel ("Incoming replies") to receive replies from workers
+
+There is no need anymore to explicitly configure the `ChunkMessageChannelItemWriter`,
+the `RemoteChunkHandlerFactoryBean` and the `MessagingTemplate` (Those can still be
+explicitly configured if required).
+
+On the worker side, the `RemoteChunkingWorkerBuilder` allows you to configure a worker to:
+
+* listen to requests sent by the master on the input channel ("Incoming requests")
+* call the `handleChunk` method of `ChunkProcessorChunkHandler` for each request
+with the configured `ItemProcessor` and `ItemWriter`
+* send replies on the output channel ("Outgoing replies") to the master
+
+There is no need anymore to explicitly configure the `ChunkProcessorChunkHandler`
+and the `AggregatorFactoryBean` (Those can still be explicitly configured if required).
+
+The following example shows how to use these APIs:
+
+[source, java]
+----
+@EnableBatchIntegration
+public class RemoteChunkingConfiguration {
+
+	@Configuration
+	public static class MasterConfiguration {
+
+		@Autowired
+		private RemoteChunkingMasterStepBuilderFactory masterStepBuilderFactory;
+
+		@Bean
+		public TaskletStep masterStep() {
+			return this.masterStepBuilderFactory.get("masterStep")
+				.chunk(100)
+				.reader(itemReader())
+				.outputChannel(requests()) // requests sent to workers
+				.inputChannel(replies())   // replies received from workers
+				.build();
+		}
+
+		// Middleware beans setup omitted
+
+	}
+
+	@Configuration
+	public static class WorkerConfiguration {
+
+		@Autowired
+		private RemoteChunkingWorkerBuilder workerBuilder;
+
+		@Bean
+		public IntegrationFlow workerFlow(DirectChannel requests, DirectChannel replies){
+			return this.workerBuilder
+				.itemProcessor(itemProcessor())
+				.itemWriter(itemWriter())
+				.inputChannel(requests) // requests received from the master
+				.outputChannel(replies) // replies sent to the master
```

Comment:
I think you need to add explain somewhere that all these channels must be configured as well.
We are in Spring Batch: that might not be so obvious for end-user what to do for Spring Integration side.

Reply:
A picture is worth thousand words: https://github.com/benas/spring-batch/blob/c6ffc5da5aec4c75fea066d7c2293c1b26182826/spring-batch-docs/asciidoc/images/remote-chunking-config.png

This is also explained in the docs.

Reply:
Why does end-user have to provide `writer`, `reader` and `processor`?
I thought in this master-worker pattern everything is done by the `RemoteChunkProcessor`...

Thanks

Reply:
OK. Never mind I see: `itemReader` is a source of data for this job - can be DB, file or any thing else.
The `itemProcessor` on the worker side is indeed the part we handle an item.

I can understand that `itemWriter` is for sinking data (e.g. DB, file etc.), but I can't understand where is the correlation with the `replies`? Do we send on our own after successful sinking?

Reply:
Well, this is the contract between the user and SB to provide at least where to read data from and where to write it (as well as the business logic of how to optionally process it). So I think this is the minimum minimorum of what the user has to provide. As a first step now, the user has also to provide SI channels and Adapters.

* For channels, we tried to create them automatically in the beginning but we faced some lifecycle issues and we decided to leave them for the user as a first step.
* For adapters, we can't take them form the user because we don't know upfront which middleware (s)he is using (jms, amqp, etc)

Reply:
> Do we send on our own after successful sinking?

The worker use the `ItemWriter` to write data directly to the sink and the `ChunkProcessorChunkHandler` to send back instances of `ChunkResponse` (see return type of `handkeChunk` method) to the master.